localstack/services/s3/models_native.py

```diff
@@ -309,6 +317,152 @@ def __init__(self, key: ObjectKey, version_id: ObjectVersionId):
+class S3Part:
+    part_number: PartNumber
+    etag: ETag
+    last_modified: datetime
+    size: int
+    checksum_algorithm: Optional[ChecksumAlgorithm]
+    checksum_value: Optional[str]
+
+    def __init__(
+        self,
+        part_number: PartNumber,
+        size: int,
+        etag: ETag,
+        checksum_algorithm: Optional[ChecksumAlgorithm] = None,
+        checksum_value: Optional[str] = None,
+    ):
+        self.last_modified = datetime.utcnow()
+        self.part_number = part_number
+        self.size = size
+        self.etag = etag
+        self.checksum_algorithm = checksum_algorithm
+        self.checksum_value = checksum_value
+
+    @property
+    def quoted_etag(self) -> str:
+        return f'"{self.etag}"'
+
+
+class S3Multipart:
+    parts: dict[PartNumber, S3Part]
+    object: S3Object
+    upload_id: MultipartUploadId
+    checksum_value: Optional[str]
+    initiated: datetime
+
+    def __init__(
+        self,
+        key: ObjectKey,
+        storage_class: StorageClass | ObjectStorageClass = StorageClass.STANDARD,
+        expires: Optional[datetime] = None,
+        expiration: Optional[datetime] = None,  # come from lifecycle
+        checksum_algorithm: Optional[ChecksumAlgorithm] = None,
+        encryption: Optional[ServerSideEncryption] = None,  # inherit bucket
+        kms_key_id: Optional[SSEKMSKeyId] = None,  # inherit bucket
+        bucket_key_enabled: bool = False,  # inherit bucket
+        lock_mode: Optional[ObjectLockMode] = None,
+        lock_legal_status: Optional[ObjectLockLegalHoldStatus] = None,
+        lock_until: Optional[datetime] = None,
+        website_redirect_location: Optional[WebsiteRedirectLocation] = None,
+        acl: Optional[str] = None,  # TODO
+        user_metadata: Optional[Metadata] = None,
+        system_metadata: Optional[Metadata] = None,
+        initiator: Optional[Owner] = None,
+    ):
+        self.id = token_urlsafe(96)  # MultipartUploadId is 128 characters long
+        self.initiated = datetime.utcnow()
+        self.parts = {}
+        self.initiator = initiator
+        self.checksum_value = None
+        self.object = S3Object(
+            key=key,
+            user_metadata=user_metadata,
+            system_metadata=system_metadata,
+            storage_class=storage_class or StorageClass.STANDARD,
+            expires=expires,
+            expiration=expiration,
+            checksum_algorithm=checksum_algorithm,
+            encryption=encryption,
+            kms_key_id=kms_key_id,
+            bucket_key_enabled=bucket_key_enabled,
+            lock_mode=lock_mode,
+            lock_legal_status=lock_legal_status,
+            lock_until=lock_until,
+            website_redirect_location=website_redirect_location,
+            acl=acl,
+        )
+
+    def complete_multipart(
+        self,
+        parts: CompletedPartList,
+        buffer: LockedSpooledTemporaryFile,
+        parts_buffers: list[LockedSpooledTemporaryFile],
+    ):
```

Comment:
Exactly, this is for now. The idea was to create an interface that all file object from a storage backend should follow, so that provider / keystore methods can have proper typing and abstraction. 
I think we add method in the storage backend, then we would to pass the multipart object to it, and we start mixing logic between storage and metadata. I liked having the provider being the entity between them, but not sure we should follow this.